Update middleware documentation to include Zenoh

aehils opened the pull request:
Fixes #5570 

## Description

<!--
Please include a summary of the changes and the related issue.
Highlight any key points or areas of concern.
-->
With the promotion of `rmw_zenoh` to Tier 1 support in ROS 2 Kilted, it became necessary to reflect this in relevant doc pages. The aim is also to provide information about the Zenoh middleware and to avoid implying DDS is the only protocol option for middleware implementations. 

The changes target these pages:

https://docs.ros.org/en/kilted/Concepts/Intermediate/About-Different-Middleware-Vendors.html
https://docs.ros.org/en/kilted/Concepts/Advanced/About-Middleware-Implementations.html

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.*
-->

### Did you use Generative AI?

<!--
If this pull request was generated using Generative AI, specify the model (e.g., GitHub Copilot v3.2)
-->
PR wasn't generated with GenAI
### Additional Information

<!--
If applicable, provide any additional context or details about the changes.
-->
- **Question:** In `About-Different-Middleware-Vendors.rst`, I added a description Zenoh in line 28.  
Should it match the shorter definition in [Non-DDS-Implementations](https://docs.ros.org/en/rolling/Installation/RMW-Implementations/Non-DDS-Implementations/Working-with-Zenoh.html), or is this a chance to improve both?
- **Question:** The docs currently don’t mention Zenoh’s inter-vendor compatibility. Are there known limitations to cross-vendor communication configurations with Zenoh? If so, I believe we should note them in `About-Different-Middleware-Vendors.rst` under the **'Cross-Vendor Communication'** subtitle (from Line 103).
> Further on this page discussed in my second question, I thought the info previously under **'Multiple RMW implementations'** about the inter-vendor communication compatibilities may have been better suited here in **'Cross-Vendor Communication'**. This is an acceptable change to make? 

🧵 While working on this, I discovered several other documentation pages with similar outdated content. I’ve created a new issue to track those additional updates: #5643.

aehils committed:
update middleware documents to include Zenoh

modifications to pages 'About middleware implementations' and 'About different middleware vendors'
- added Zenoh information and fixed links
- adjusted wording to avoid implying DDS is the only protocol option for middleware implementations

aehils committed:
Fix linting issues in files to update middleware docs

fixes in both About-Middleware-Implementations.rst and
About-Different-Middleware-Vendors.rst

fujitatomoya commented on source/Concepts/Advanced/About-Middleware-Implementations.rst:
```diff
@@ -48,14 +50,36 @@ This is done using the ``.idl`` files generated by the ``rosidl_generator_dds_id
+Unlike DDS-based implementations, this middleware relies on a Zenoh router to discover peers and pass discovery information along via Zenoh's 'gossip scouting'.
+Therefore, ``rmw_zenoh_cpp`` requires the Zenoh router (``zenohd``) to be active on the local system or reachable over the network.
+
+In ROS 2's Zenoh integration, each `context <https://docs.ros.org/en/rolling/p/rclcpp/generated/classrclcpp_1_1Context.html#:~:text=Context%20which%20encapsulates%20shared%20state,init%20and%20shutdown%20of%20rclcpp.>`_ is mapped to a single Zenoh session.
```
```suggestion
In ROS 2's Zenoh integration, each `context <https://docs.ros.org/en/{DISTRO}/p/rclcpp/generated/classrclcpp_1_1Context.html>`_ is mapped to a single Zenoh session.
```

fujitatomoya commented on source/Concepts/Advanced/About-Middleware-Implementations.rst:
```diff
@@ -48,14 +50,36 @@ This is done using the ``.idl`` files generated by the ``rosidl_generator_dds_id
+Therefore, ``rmw_zenoh_cpp`` requires the Zenoh router (``zenohd``) to be active on the local system or reachable over the network.
+
+In ROS 2's Zenoh integration, each `context <https://docs.ros.org/en/rolling/p/rclcpp/generated/classrclcpp_1_1Context.html#:~:text=Context%20which%20encapsulates%20shared%20state,init%20and%20shutdown%20of%20rclcpp.>`_ is mapped to a single Zenoh session.
+This session is shared across all publishers, subscriptions, services, and clients within that context.
+The context maintains a local graph cache that tracks the network topology of ROS 2 entities and the presence of each entity is managed through unique liveliness tokens issued on creation and revoked during destruction.
```
if we want to add this information to zenoh, why not adding the same information to DDS based rmw implementations above? maybe we can add the explanation about DDS discovery listener, DDS participant that is mapped to the context and so on?

or maybe this is overkill for here? probably we just want to share the overview about rmw_zenoh_cpp here? if that is so, we can skip the following details from here? (possibly we can add these details for https://github.com/ros2/rmw_zenoh)

aehils replied:
i added this here, just as you said, to give an overview of the structure behind how zenoh works in ROS 2. i though it might suffice as a complement to the information provided in this page about the DDS packages.

perhaps it may be overkill to include that additional information here? however, i trust your judgement. if you believe this is a good place for that info on DDS rmw, i will be happy to put that together and add to section above.

also, im happy you pointed out `rmw_zenoh`, that info is actually available in the repo in `rmw_zenoh/docs/design.md`

fujitatomoya replied:
okay, let's keep it here. if we think we want to move this to zenoh doc, we can always change that.

fujitatomoya commented on source/Concepts/Advanced/About-Middleware-Implementations.rst:
```diff
@@ -48,14 +50,36 @@ This is done using the ``.idl`` files generated by the ``rosidl_generator_dds_id
+Here is an inexhaustive list of how the Zenoh middleware |API| adapts ROS 2 entities over its communication protocol:
 
-As an example of an rmw implementation repository, the ``Eclipse Cyclone DDS`` ROS middleware implementation is on |GitHub|_ at `ros2/rmw_cyclonedds <https://github.com/ros2/rmw_cyclonedds>`_.
+- **Nodes:** Nodes in ROS 2 have been referred to as "units of computation" in a ROS 2 graph, whereby each node should be responsible for a single, modular purpose. Zenoh has no direct counterpart to the node, so ``rmw_zenoh_cpp`` creates no Zenoh entities for them. However, when a node is created through the RMW |API|, a liveliness token of type ``NN`` is declared.
```
i believe that github action fails on this line and else where, because of `one-sentence-per-line` rule.

aehils replied:
yes, i was having issues with lint here. it was either i broke the `one-sentence-per-line` rule, or the list formatting returned errors. i suppose the solution here will be to add a blank line between list items.

aehils replied:
i have use backslashes instead, hoping that allows me to obey the `one-sentence-per-line` rule and maintain the list structure

aehils replied:
the backslashes didnt help. i resorted to removing the bullets in the list altogether, and listing my points here as separate paragraphs

fujitatomoya commented on source/Concepts/Intermediate/About-Different-Middleware-Vendors.rst:
```diff
@@ -46,43 +54,51 @@ Supported RMW implementations
 ----------------------------
 
 The ROS 2 binary releases for currently active distros have built-in support for several RMW implementations out of the box (Fast DDS, RTI Connext Pro, Eclipse Cyclone DDS, GurumNetworks GurumDDS).
+Beginning with ROS 2 Kilted, this also includes Eclipse Zenoh.
```
maybe it can be precise.

```suggestion
Beginning with ROS 2 Kilted Kaiju, this also includes Eclipse Zenoh.
```

fujitatomoya commented on source/Concepts/Intermediate/About-Different-Middleware-Vendors.rst:
```diff
@@ -92,5 +108,13 @@ See the :doc:`guide <../../How-To-Guides/Working-with-multiple-RMW-implementatio
 --------------------------
 
+For many cases you will find that nodes using different RMW implementations are able to communicate, however this is not true under all circumstances.
+Here is a list of inter-vendor communication configurations that are not supported:
+
+- Fast DDS <-> Connext
+   - ``WString`` published by Fast DDS can't be received correctly by Connext on macOS
+- Connext <-> Cyclone DDS
+   - does not support pub/sub communication for ``WString``
```
i cannot say there are the only condition, i think service does not work cross-vendor communication, that means action does not work either. (there can be more reasons and conditions as it is being developed.)
i would avoid listing unsupported communication details here, instead it can just say that is not officially supported as it is?

aehils replied:
got it. i had retained those details from the previous version of the page, but i will remove that part to maintain focus on the lack of support for cross-vendor communication generally.

fujitatomoya commented on source/Concepts/Intermediate/About-Different-Middleware-Vendors.rst:
```diff
@@ -92,5 +108,13 @@ See the :doc:`guide <../../How-To-Guides/Working-with-multiple-RMW-implementatio
 Cross-Vendor Communication
 --------------------------
 
+For many cases you will find that nodes using different RMW implementations are able to communicate, however this is not true under all circumstances.
```
1st of all, i think we should mention that this section is for DDS based RMW implementations? it is obvious that rmw_zenoh and rmw_cyclonedds does not communication?

aehils replied:
right okay, i will make it clear that this section talks about cross-vendor communication among DDS implementations.

fujitatomoya reviewed (commented):
Overall, i think this is great enhancement and update. i have several comments.

aehils committed:
doc fix: refine updates to middleware docs

- clarify focus of 'Cross-Vendor Communication' on DDS
- removed unsupported configurations listed in 'Cross-Vendor
Communication' to address lack of support for inter-vendor generally
- more precise naming for ROS 2 Kilted Kaiju, not just 'ROS 2 Kilted'
- fixed lint issues related to bullet list by removing the list
altogether, items now written as separate paragraphs

fujitatomoya commented on source/Concepts/Advanced/About-Middleware-Implementations.rst:
```diff
@@ -48,14 +50,50 @@ This is done using the ``.idl`` files generated by the ``rosidl_generator_dds_id
+Unlike DDS-based implementations, this middleware relies on a Zenoh router to discover peers and pass discovery information along via Zenoh's 'gossip scouting'.
+Therefore, ``rmw_zenoh_cpp`` requires the Zenoh router (``zenohd``) to be active on the local system or reachable over the network.
+
+In ROS 2's Zenoh integration, each `context <https://docs.ros.org/en/rolling/p/rclcpp/generated/classrclcpp_1_1Context.html>`_ is mapped to a single Zenoh session.
```
`{DISTRO}` will be relaced with `rolling` and other distro names during sphinx build.

```suggestion
In ROS 2's Zenoh integration, each `context <https://docs.ros.org/en/{DISTRO}/p/rclcpp/generated/classrclcpp_1_1Context.html>`_ is mapped to a single Zenoh session.
```

fujitatomoya commented:
@aehils can you rebase this, it says there are conflicts against rolling. besides that i have only minor comment to be fixed.

aehils commented:
hey @fujitatomoya i made that last change to the link. i didnt realise the URL would work that way, thanks for letting me know!

i also rebased the branch now, looks like the merge conflicts are sorted.

aehils committed:
fix linting issues in source/Concepts/Advanced/About-Middleware-Implementations.rst

aehils commented:
sorry @fujitatomoya, i had to fix linting issues again. looks good to go now. thanks!!

aehils committed:
fix missing reference 'different-middleware-vendors-cross-vendor-communication'

fujitatomoya commented on source/Concepts/Intermediate/About-Different-Middleware-Vendors.rst:
```diff
@@ -91,8 +107,9 @@ See the :doc:`guide <../../How-To-Guides/Working-with-multiple-RMW-implementatio
 
 .. _different-middleware-vendors-cross-vendor-communication:
```
@aehils for https://github.com/ros2/ros2_documentation/issues/5643#issuecomment-2905206278, this is what i mean.
